src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs

```diff
@@ -40,8 +40,47 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
             | nameof(x)       |      |       |             |             |       ✔️        | ️
             | sizeof(x)       |      |       |             |             |       ✔️        | ️
             | typeof(x)       |      |       |             |             |       ✔️        | ️
+            | out var x       |      |  ✔️   |             |             |                 | ️
+            | case X x:       |      |  ✔️   |             |             |                 | ️
+            | obj is X x      |  ✔️  |  ✔️   |             |             |                 |
 
             */
+            if (operation is ILocalReferenceOperation localReference &&
+                localReference.IsDeclaration &&
+                !localReference.IsImplicit) // Workaround for https://github.com/dotnet/roslyn/issues/30753
+            {
+                // Declaration expression is a definition (write) for the declared local.
+                return ValueUsageInfo.Write;
+            }
+            else if (operation is IDeclarationPatternOperation)
+            {
+                switch (operation.Parent)
+                {
+                    case IPatternCaseClauseOperation _:
+                        // A declaration pattern within a pattern case clause is a
+                        // write for the declared local.
+                        // For example, 'x' is defined and assigned the value from 'obj' below:
+                        //      switch (obj)
+                        //      {
+                        //          case X x:
+                        //
+                        return ValueUsageInfo.Write;
+
+                    case IIsPatternOperation _:
+                        // A declaration pattern within an is pattern is a
+                        // both a write and read for the declared local.
+                        // For example, 'x' is both written and read in the following expression:
+                        //      if (obj is X x)
+                        //
+                        return ValueUsageInfo.ReadWrite;
```

Comment:
Isn't there a null check here, i.e. if condition is true if `x != null`. I might have misinterpreted the language here if that is not the case...

---
In reply to: [231664635](https://github.com/dotnet/roslyn/pull/31025#discussion_r231664635) [](ancestors = 231664635)

Reply:
1.  interesting.  i hadn't considered that interpretation.  I would have thought of 'obj' as read, but not 'x' itself.  I'm not sure hte best way to think about that.  If that is the interpretation, it might be good to doc that.
2. there will be a 'read' here for the `is X x` case, but not for the `is var x` case.  Is that a different IOperation?

Reply:
> If that is the interpretation, it might be good to doc that.

Will do.

> 2. there will be a 'read' here for the is X x case, but not for the is var x case.  Is that a different IOperation?

Yes, that generates a local reference operation with `IsDeclaration` flag set to true. That is the case handle in the topmost if added here, and we consider it as a Write-only operation.

---
In reply to: [232104206](https://github.com/dotnet/roslyn/pull/31025#discussion_r232104206) [](ancestors = 232104206)

Reply:
Oh interesting!  so `is var x` isn't even a pattern in the IOp world?

Reply:
> Yes, that generates a local reference operation with IsDeclaration flag set to true. That is the case handle in the topmost if added here, and we consider it as a Write-only operation.

Cool.  If you have tests for tehse cases, them i'm good with this approach :)  Thanks for the explanations!

Reply:
Actually, I might have misinterpreted your question. Both of the cases you mentioned would be identical operations. I think we should be treating them as write only, and if there is no use of x, it is probably an unused value that can be replaced with discard. Let me also add a test.

---
In reply to: [232105584](https://github.com/dotnet/roslyn/pull/31025#discussion_r232105584) [](ancestors = 232105584,232104206)

Reply:
I'm glad i brought it up then! :D

Reply:
Fixed with unit tests in [https://github.com/dotnet/roslyn/pull/30777/commits/2d619ae9318c1c5b502827acd66166aec768b9e2](https://github.com/dotnet/roslyn/pull/30777/commits/2d619ae9318c1c5b502827acd66166aec768b9e2). Also helped delete bunch of code in the walker that was handling the side effects of treating it as a read write. I will push similar change on this branch (without unit tests) as well.

---
In reply to: [232107149](https://github.com/dotnet/roslyn/pull/31025#discussion_r232107149) [](ancestors = 232107149)